src/services/staking_vaults.py

```diff
@@ -135,54 +119,140 @@ def get_vaults_total_values(
+        ]
+
+        pubkeys_to_check = non_eligible_to_activation_validators_pubkeys + deposit_pubkeys
+        if not pubkeys_to_check:
+            return {}
+
         return self.w3.lido_contracts.lazy_oracle.get_validator_statuses(
-            pubkeys=pubkeys,
+            pubkeys=pubkeys_to_check,
             block_identifier=block_identifier,
+            batch_size=variables.VAULT_VALIDATOR_STATUSES_BATCH_SIZE,
         )
 
-    @staticmethod
-    def _get_total_pending_amount_by_pubkey(
-        pending_deposits: list[PendingDeposit],
-    ) -> dict[str, Gwei]:
-        deposits: defaultdict[str, int] = defaultdict(int)
-        for deposit in pending_deposits:
-            deposits[deposit.pubkey] += int(deposit.amount)
-
-        return {pubkey: Gwei(deposits[pubkey]) for pubkey in deposits}
-
-    @staticmethod
-    def _get_validators_by_vault(validators: list[Validator], vaults: VaultsMap) -> VaultToValidators:
+    def _calculate_vault_total_value(
+        self,
+        vault_address: str,
+        vault_aggregated_balance: Wei,
+        vault_validators: list[Validator],
+        vault_pending_deposits: list[PendingDeposit],
+        total_pending_amount_by_pubkey: dict[str, Gwei],
+        validator_statuses: dict[str, ValidatorStatus],
+    ) -> int:
         """
-        Groups validators by their associated vault, based on withdrawal credentials.
+        Calculates total value for a single vault.
         """
-        wc_to_vault: dict[str, VaultInfo] = {v.withdrawal_credentials: v for v in vaults.values()}
+        vault_total = int(vault_aggregated_balance)
+
+        for validator in vault_validators:
+            validator_pubkey = validator.validator.pubkey
+            validator_pending_amount = total_pending_amount_by_pubkey.get(validator_pubkey, Gwei(0))
+            total_validator_balance = gwei_to_wei(Gwei(validator.balance + validator_pending_amount))
+
+            # Include validator balance and all pending deposits in TV when validator is eligible for activation or
+            # has already passed activation
+            if not has_far_future_activation_eligibility_epoch(validator.validator):
+                vault_total += int(total_validator_balance)
+
+            # For not-yet-eligible validators, use PDG stages:
+            # - PREDEPOSITED: add 1 ETH (guaranteed)
+            # - ACTIVATED: add full balance + pending deposits
+            # All other stages are skipped as not related to the non-eligible for activation validators
+            else:
+                status = validator_statuses.get(validator_pubkey)
+                # Skip if validator pubkey in PDG is not associated with the current vault
+                if status is None or status.staking_vault != vault_address:
+                    continue
+
+                if status.stage == ValidatorStage.PREDEPOSITED:
+                    vault_total += int(gwei_to_wei(MIN_DEPOSIT_AMOUNT))
```

Comment:
What if validator already has pending deposit in 31 ETH. Should we add it already here to check if balance > 32 ETH?